### DISCUSSION TITLE
How do I update one item with the other item's metadata?

### DISCUSSION THREAD

**Discussion by Nirmal4G** [dotnet/msbuild#7753] [Q&A]
The more elaborate question would be...

How do I update one item with the other item's metadata when their identities are different and somehow I need to compare different metadata for equality?

```xml
  <Target Name="Default">
    <ItemGroup>
      <A Update="@(B->'%(Meta0)')" Meta0="%(B.Identity)" Meta1="%(B.Meta1)%(B.Meta2)" />
    </ItemGroup>
  </Target>
```

For all the `B.Meta0` items present in `A` items, update `A.Meta0` with `B`'s Identity and `A.Meta1` with `B.Meta{1,2}`

**Comment by KalleOlaviNiemitalo** [comment]
According to [Item element (MSBuild)](https://docs.microsoft.com/en-us/visualstudio/msbuild/item-element-msbuild?view=vs-2022), the Update attribute is not valid in an Item in an ItemGroup in a Target. I wish MSBuild warned about that.

So, if you need to do this in a Target and cannot use Update, I think a custom task (inline or otherwise) will be the easiest solution.

**Reply by Nirmal4G** [reply]
I know but I also know there's another way using batching but I myself don't know how to write it correctly. That's why I asked here.

> I wish MSBuild warned about that (_not supporting the Update on the item in the target_).

Me Too!

**Reply by rainersigwald** [reply]
Me three, but it is a breaking change that we have to consider fairly carefully: dotnet/msbuild#2835

**Reply by Nirmal4G** [reply]
Let me think about this out loud, previously specifying `Update` is ignored and processed all items, meaning that anyone who used Update should know that their code wasn't working, like me and must have used batching to solve this. So, even if we provide `Update` filter does the batching still happen or do we go through different route for `Update` as in regular item groups?

For Example

```xml
<!-- V1 -->
<Item1 Condition="'%(Identity)' != '' AND '@(Item1)' != '' AND '@(Item2)' != ''" Metadata="Updated" />

<!-- V2 -->
<Item1 Update="@(Item2)" Condition="'%(Identity)' != '' AND '@(Item1)' != '' AND '@(Item2)' != ''" Metadata="Updated" />

<!-- V3 -->
<Item1 Update="@(Item2)" Metadata="Updated" />
```

Users could've written V1, V2 but unlikely V3 when Update doesn't work like they intended. V1 and V2 are similar now but that'll change when `Update` is supported. Hence, the breaking change. But does it also mean that actual user code still produces the same output even after `Update` is supported since both filters are same?

**So, what's the actual difference between the above three variants which is preventing us from supporting `Update` in items inside targets?**

**Comment by KalleOlaviNiemitalo** [comment]
For batching, this seems to work:

```XML
<?xml version="1.0" encoding="utf-8"?>
<!-- https://github.com/dotnet/msbuild/discussions/7753 -->
<Project>
  <Target Name="Default">
    <ItemGroup>
      <A Include="Blue" Meta0="Carrot" Meta1="Happy" Meta2="Pitch" />
      <A Include="Red" Meta0="Bilberry" Meta1="Anxious" Meta2="Yaw" />
      <A Include="Cyan" Meta0="Rhubarb" Meta1="Tired" Meta2="Roll" />
    </ItemGroup>
    <ItemGroup>
      <B Include="Garlic" Meta0="Blue" Meta1="Curious" Meta2="Desk" />
      <B Include="Cucumber" Meta0="Red" Meta1="Flabbergasted" Meta2="Chair" />
      <B Include="Olive" Meta0="Purple" Meta1="Clever" Meta2="Bookshelf" />
    </ItemGroup>

    <ItemGroup>
      <_SwapB Include="%(B.Meta0)">
        <Meta0>%(Identity)</Meta0>
        <Meta1>%(Meta1)%(Meta2)</Meta1>
      </_SwapB>
      <_ModifiedA Include="@(A)" Condition="%(Identity) != '' and '@(_SwapB)' != ''">
        <Meta0>@(_SwapB->'%(Meta0)')</Meta0>
        <Meta1>@(_SwapB->'%(Meta1)')</Meta1>
      </_ModifiedA>
      <A Remove="@(_ModifiedA)" />
      <A Include="@(_ModifiedA)" />
    </ItemGroup>

    <Message Importance="high" Text="A: Identity=%(A.Identity) Meta0=%(A.Meta0) Meta1=%(A.Meta1) Meta2=%(A.Meta2)" />
  </Target>
</Project>
```

Output:

```
  A: Identity=Cyan Meta0=Rhubarb Meta1=Tired Meta2=Roll
  A: Identity=Blue Meta0=Garlic Meta1=CuriousDesk Meta2=Pitch
  A: Identity=Red Meta0=Cucumber Meta1=FlabbergastedChair Meta2=Yaw
```

**Reply by KalleOlaviNiemitalo** [reply]
However, if you need to preserve the order of `A` items, then I don't know how to do that.

**Reply by Nirmal4G** [reply]
While looking into your solution, I also found a way to make this even simpler and as a side effect, it seems to preserve order at lease in my tests but do please test and confirm before using it!

**Comment by Nirmal4G** [comment]
Here's another (_simpler_) variant of the solution proposed by @KalleOlaviNiemitalo

```xml
<ItemGroup>
	<_SwapB Include="@(B->'%(Meta0)')">
		<Meta0>%(B.Identity)</Meta0>
		<Meta1>%(B.Meta1)%(B.Meta2)</Meta1>
		<Meta2/> <!-- this not needed it seems but I don't know why! -->
	</_SwapB>

	<A Condition="'%(Identity)' != '' AND '@(A)' != '' AND '@(_SwapB)' != ''">
		<Meta0>@(_SwapB->'%(Meta0)')</Meta0>
		<Meta1>@(_SwapB->'%(Meta1)')</Meta1>
	</A>

	<_SwapB Remove="@(_SwapB)"/>
</ItemGroup>
```

This is almost the same except we don't need to Modify `A` items separately. I still don't like the extra `_SwapB` but I don't know how to optimize any more than this!

But a request to MSBuild team... @Forgind @rainersigwald
I would certainly like my proposed code better than this.
Please provide `Update` support for items within targets!
